internal/graph/check_test.go

```diff
@@ -119,3 +121,203 @@ func TestCheckWithOneConcurrentGoroutineCausesNoDeadlock(t *testing.T) {
+			check:            tuple.NewTupleKey("document:x", "a", "user:unknown"),
+			contextualTuples: []*openfgav1.TupleKey{tuple.NewTupleKey("document:x", "a", "user:unknown")},
+			allowed:          true,
+			minDBReads:       1, // checkDirectUserTuple needs to run
+			maxDBReads:       2, // maybe checkDirectUsersetTuples also runs
+		},
+		{
+			name:       "union",
+			check:      tuple.NewTupleKey("document:x", "union", "user:maria"),
+			allowed:    true,
+			minDBReads: 1, // checkDirectUserTuple needs to run
+			maxDBReads: 2, // maybe checkDirectUsersetTuples also runs
+		},
+		{
+			name:       "union_no_access",
+			check:      tuple.NewTupleKey("document:x", "union", "user:unknown"),
+			allowed:    false,
+			minDBReads: 4, // need to check all the conditions in the union, so two direct tuple lookups + two userset lookups
+			maxDBReads: 4,
+		},
+		{
+			name:       "intersection",
+			check:      tuple.NewTupleKey("document:x", "intersection", "user:maria"),
+			allowed:    true,
+			minDBReads: 2, // need at minimum two direct tuple checks
+			maxDBReads: 4, // at most two tuple checks + two userset checks
+		},
+		{
+			name:       "intersection_no_access",
+			check:      tuple.NewTupleKey("document:x", "intersection", "user:unknown"),
+			allowed:    false,
+			minDBReads: 2, // need at minimum two direct tuple checks
+			maxDBReads: 4, // at most two tuple checks + two userset checks
+		},
+		{
+			name:       "difference",
+			check:      tuple.NewTupleKey("document:x", "difference", "user:jon"),
+			allowed:    true,
+			minDBReads: 2, // need at minimum two direct tuple checks
+			maxDBReads: 4, // at most two tuple checks + two userset checks
+		},
+		{
+			name:       "difference_no_access",
+			check:      tuple.NewTupleKey("document:x", "difference", "user:maria"),
+			allowed:    false,
+			minDBReads: 1, // if the "but not" condition returns quickly with "false", no need to evaluate the first branch
+			maxDBReads: 4, // at most two tuple checks + two userset checks
+		},
+		{
+			name:       "ttu",
+			check:      tuple.NewTupleKey("document:x", "ttu", "user:maria"),
+			allowed:    true,
+			minDBReads: 2, // one read to find org:fga + one direct check if user:maria is a member of org:fga
+			maxDBReads: 3, // one read to find org:fga + (one direct check + userset check) if user:maria is a member of org:fga
+		},
+		{
+			name:       "ttu_no_access",
+			check:      tuple.NewTupleKey("document:x", "ttu", "user:jon"),
+			allowed:    false,
+			minDBReads: 3, // one read to find org:fga + (one direct check + userset check) to see if user:jon is a member of org:fga
+			maxDBReads: 3,
+		},
+		// more complex scenarios
+		{
+			name:       "union_and_ttu",
+			check:      tuple.NewTupleKey("document:x", "union_and_ttu", "user:maria"),
+			allowed:    true,
+			minDBReads: 3, // union (1 read) + ttu (2 reads)
+			maxDBReads: 5, // union (2 reads) + ttu (3 reads)
```

Comment:
This should be 7, worst case.

union decomposes into `a or b`
* 2 reads worst case for `document#a`
* 2 reads worst case for `document#b`

`document#ttu` has worst case of 3 reads (1 for tupleset lookup + 2 for direct and userset lookups)

4 reads worst case for `document#union` + 3 reads worst case for `document#ttu` is a total of 7 reads.

Reply:
🤔 wait, I'm probably wrong because it's intersection. Let me second guess myself..

Reply:
Ok, correction..

This should be `max(max(union), max(ttu))` 

`max(union) == 4`
`max(ttu) == 3`

So this should be 4, not 5.

Reply:
Ok, I stand corrected.

Shouldn't this be `max(max(union), max(ttu))`?

If the union evaluates first it could take up to `4`, but if the ttu evaluates first it could take up to `3`. The worst case of the two is due to the union case.

Reply:
Oh, nvm. I see where you got this from.

`a or b` will resolve in  a max of 2 because `user:maria` has direct access, so one of the paths is redundant. Then we have to evaluate the TTU (because of the intersection) which will produce a up to a maximum of 3 additional reads. So 2 + 3 = 5.

Gotcha 👍